### PR TITLE
[CP] Add Ulysses CP for varlen attention

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #4459

Share Ulysses resharding between FlexAttention and VarlenAttention. Add a varlen integration recipe and reject Ulysses kernels for GPT-OSS, whose TP-only sink sharding is incompatible.

Tests: Add CPU coverage and a 2-GPU CP integration case. Over 10 deterministic steps, flex and varlen differ by at most 7.534e-05 in loss and 5.960e-05 in grad norm.

### CODE REVIEW

File: torchtitan/models/common/cp_attention.py [hunk 1]
```diff
@@ -164,6 +162,24 @@ def forward(
...
         out_THV = super().forward(q_THK, k_THK, v_THV, **kwargs)
         # Back to sharded tokens: (T, H/cp, V) -> (T/cp, H, V).
         return self._reshard(out_THV, cp_group, src=_HEAD_DIM, dst=_SEQ_DIM)
+
+
+class UlyssesCPFlexInnerAttention(UlyssesCPInnerAttention, FlexInnerAttention):
+    """FlexInnerAttention under Ulysses CP."""
+
+    @dataclass(kw_only=True, slots=True)
+    class Config(FlexInnerAttention.Config):
```
Comment (tianyu-l) [inline torchtitan/models/common/cp_attention.py:181]: can we inherit both config, so that we don't need to rely on owner in gpt_oss/parallelize.py? Not sure if there's anything with multi-inheritance.

File: torchtitan/models/common/cp_attention.py [hunk 2]
```diff
@@ -164,6 +162,24 @@ def forward(
...
+        pass
+
+
+class UlyssesCPVarlenInnerAttention(UlyssesCPInnerAttention, VarlenInnerAttention):
+    """VarlenInnerAttention under Ulysses CP."""
+
+    @dataclass(kw_only=True, slots=True)
+    class Config(VarlenInnerAttention.Config):
```
Comment (tianyu-l) [inline torchtitan/models/common/cp_attention.py:189]: similar

File: torchtitan/models/common/cp_attention.py [hunk 3]
```diff
@@ -164,6 +162,24 @@ def forward(
             self._reshard(x, cp_group, src=_SEQ_DIM, dst=_HEAD_DIM)
             for x in (q_THK, k_THK, v_THV)
         )
+        # The concrete subclass provides the attention implementation.
+        # pyrefly: ignore [missing-attribute]
         out_THV = super().forward(q_THK, k_THK, v_THV, **kwargs)
```
Comment (tianyu-l) [inline torchtitan/models/common/cp_attention.py:167]: wait, how does it figure out that the call chain should be
UlyssesCPInnerAttention (this) -> Flex/VarlenInnerAttention (super)?

File: torchtitan/models/common/cp_attention.py (hunk 1 above)
Comment (fegin) [reply torchtitan/models/common/cp_attention.py:181]: `UlyssesCPInnerAttention` doesn't have the configuration at this moment. So I think it is better to not do this just to avoid the `_owner` check . I change how we check in `gpt_oss/parallelize.py` so that we don't use _owner private field.

File: torchtitan/models/common/cp_attention.py (hunk 3 above)
Comment (fegin) [reply torchtitan/models/common/cp_attention.py:167]: `super()` continues along the class's MRO. Since `CPInnerAttention` and `ABC` do not define `forward()`, it reaches `FlexInnerAttention.forward()` or `VarlenInnerAttention.forward()`.

File: torchtitan/models/common/cp_attention.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/models/common/cp_attention.py:181]: we do it all the time, see
- https://github.com/pytorch/torchtitan/blob/91abd23019bc64a0990e955411112f1fa4466320/torchtitan/components/data/collators.py#L57
- https://github.com/pytorch/torchtitan/blob/91abd23019bc64a0990e955411112f1fa4466320/torchtitan/protocols/state_dict_adapter.py#L38
- https://github.com/pytorch/torchtitan/blob/91abd23019bc64a0990e955411112f1fa4466320/torchtitan/components/tokenizer.py#L115
- https://github.com/pytorch/torchtitan/blob/91abd23019bc64a0990e955411112f1fa4466320/torchtitan/models/llama3/model.py#L35

In fact I think this should be the preferred way of coding, because the module you invent, say `A`, will likely be inherited by other people e.g. `B(A)`, who would take for granted and do `B.Config(A.Config)`, only to find that `A.Config` doesn't exist.

Comment (fegin) [reply torchtitan/models/common/cp_attention.py]: okay, the inheritance argument is strong. I'll change it.

**Review by tianyu-l — APPROVED** [review]
